identity-flow/design.md

```diff
@@ -0,0 +1,353 @@
+`FlowException` annotation and registered with `FlowDispatcher` on both server and the client,
+they will be caught, marshalled, and rethrown from the client stub. By convention, when operating
+through HTTP uncaught `UnsupportedOperationException` is mapped to HTTP status 404 and
+uncaught `IllegalStateException` is mapped to HTTP status 405. All other uncaught exceptions are
+mapped to HTTP status 500.
+
+## Dispatcher
+
+The lowest level of the Flow RPC stack is a dispatcher. It is represented by this interface:
+
+```kotlin
+interface FlowDispatcher {
+    val exceptionMap: FlowExceptionMap
+    suspend fun dispatch(flow: String, method: String, args: List<DataItem>): List<DataItem>
+}
+```
+
+Here `flow` is the path of the flow, `method` is method name to be called and `args`
+represents flow state and method arguments. The first element is always opaque server-side
+state: Cbor-serialized state object encrypted with the secret server key and wrapped into
+Cbor `Bstr`. Other elements of the array encode method arguments one by one. For data arguments
+their Cbor representation is used. Flow arguments are encoded as two-element Cbor array, where
+the first element is flow path (as `Tstr`) and the second is opaque flow state (as `Bstr`).
+
+Result array is encoded as three- or four- element array: the first element is updated opaque
+flow state, the second element is an integer indicating result type: `0` for normal result
+and `1` for an exception. For the normal result the third element represents the returned value.
+For an exception, the third element is an exception id, and the fourth is Cbor representation
+of the exception object.
+
+Kotlin `Unit` return type is represented as an empty `Bstr`. Data return types and exceptions
+are represented by their Cbor representations. Returned flows are represented as three-element
+arrays. The first element of the array is a `Tstr` that holds result flow's path, the second is
+the name of the flow _join_ method as `Tstr` (see the next section) and the third is the opaque
+state of the result flow.
+
+## Flow creating and joining
+
+A from which is marked with `creatable = true` in its `FlowState` annotation, can be created
+directly by the client. By default the stub implementation class generated by the annotation
+processor from the interface has the same name as the interface with `Impl` suffix added at the
+end. Stub implementation class constructor will be defined like this:
+
+```kotlin
+class ExampleImpl(flowPath: String, flowState: DataItem,
+    flowDispatcher: FlowDispatcher, flowNotifier: FlowNotifier,
+    onComplete: suspend (DataItem) -> Unit = {})
+```
+
+By convention, empty `Bstr` passed to `flowState` parameter will be translated into creating
+a new state object that corresponds to the given `flowPath` with no parameters. `FlowNotifier`
+will be covered in a later section.
+
+A common pattern for an API on a flow interface `A` to create another flow interface `B`, with the
```

Comment:
"A common pattern for an API" -> "A common pattern is for an API"

Reply:
Done